File: sdk/ctv/android-tv/src/main/kotlin/com/rivalapexmediation/ctv/metrics/MetricsRecorder.kt
package com.rivalapexmediation.ctv.metrics

import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import com.rivalapexmediation.ctv.ApexMediation
import com.rivalapexmediation.ctv.SDKConfig
import com.rivalapexmediation.ctv.network.LoadError
import com.rivalapexmediation.ctv.network.reason
import com.rivalapexmediation.ctv.render.VideoProgressEvent
import com.rivalapexmediation.ctv.util.Logger
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.util.Collections
import java.util.Locale
import java.util.concurrent.ConcurrentHashMap

internal object MetricsRecorder {
    private const val FLUSH_INTERVAL_MS = 30_000L
    private const val MAX_LATENCY_SAMPLES = 200
    private val handler = Handler(Looper.getMainLooper())
    private val counters = ConcurrentHashMap<String, Long>()
    private val latencies = Collections.synchronizedList(mutableListOf<Long>())
    private val client = OkHttpClient.Builder().build()
    private var initialized = false
    private lateinit var cfg: SDKConfig
    private var flushScheduled = false
    private val flushRunnable = Runnable {
        flush()
        flushScheduled = false
    }

    fun initialize(config: SDKConfig) {
        cfg = config
        initialized = true
    }

    fun recordRequest(durationMs: Long, error: LoadError?) {
        if (!shouldCollect()) return
        increment("requests_total")
        val success = error == null || error is LoadError.NoFill
        if (success) increment("requests_success") else increment("requests_error")
        val reasonKey = error?.reason()
        if (!reasonKey.isNullOrBlank()) increment("requests_error_$reasonKey")
        addLatency(durationMs)
        scheduleFlush()
    }

    fun recordPlayback(event: VideoProgressEvent) {
        if (!shouldCollect()) return
        val key = "playback_${event.name.lowercase(Locale.ROOT)}"
        increment(key)
        scheduleFlush()
    }

    fun recordTracker(event: String, success: Boolean) {
        if (!shouldCollect() || event.isEmpty()) return
        val key = "tracker_${event.lowercase(Locale.ROOT)}_${if (success) "success" else "failure"}"
        increment(key)
        scheduleFlush()
    }

    private fun addLatency(durationMs: Long) {
        if (durationMs <= 0) return
        synchronized(latencies) {
            if (latencies.size >= MAX_LATENCY_SAMPLES) {
                latencies.removeAt(0)
            }
            latencies.add(durationMs)
        }
    }

    private fun increment(key: String, delta: Long = 1) {
        counters.merge(key, delta) { old, add -> old + add }
    }

    private fun scheduleFlush() {
        if (flushScheduled) return
        flushScheduled = true
        handler.postDelayed(flushRunnable, FLUSH_INTERVAL_MS)
    }

    private fun flush() {
        if (!shouldCollect()) {
            counters.clear()
            synchronized(latencies) { latencies.clear() }
            return
        }
        val snapshot = buildPayload() ?: return
        val body = snapshot.toString().toRequestBody("application/json".toMediaType())
        val url = cfg.apiBaseUrl.trimEnd('/') + "/sdk/metrics"
        val builder = Request.Builder().url(url).post(body).header("Content-Type", "application/json")
        cfg.apiKey?.let { builder.header("Authorization", "Bearer $it") }
        client.newCall(builder.build()).enqueue(object: okhttp3.Callback {
            override fun onFailure(call: okhttp3.Call, e: java.io.IOException) {
                Logger.w("CTV metrics post failed: ${e.message}", e)
            }
            override fun onResponse(call: okhttp3.Call, response: okhttp3.Response) {
                response.close()
            }
        })
        counters.clear()
        synchronized(latencies) { latencies.clear() }
    }

    private fun buildPayload(): JSONObject? {
        if (counters.isEmpty() && latencies.isEmpty()) return null
        val root = JSONObject()
        val countersJson = JSONObject()
        counters.forEach { (key, value) -> countersJson.put(key, value) }
        root.put("counters", countersJson)
        if (latencies.isNotEmpty()) {
            val sorted = synchronized(latencies) { latencies.sorted() }
            val percentiles = JSONObject()
            percentiles.put("p50", percentile(sorted, 0.50))
            percentiles.put("p95", percentile(sorted, 0.95))
            percentiles.put("p99", percentile(sorted, 0.99))
            root.put("request_latency_ms", percentiles)
        }
        root.put("timestamp", System.currentTimeMillis())
        root.put("sdk", "android_tv")
        root.put("appId", cfg.appId)
        return root
    }

    private fun percentile(values: List<Long>, fraction: Double): Long {
        if (values.isEmpty()) return 0
        val index = ((values.size - 1) * fraction).toInt().coerceIn(0, values.size - 1)
        return values[index]
    }

    private fun shouldCollect(): Boolean = initialized && ApexMediation.metricsEnabled()
}

File: sdk/ctv/tvos/Sources/CTVSDK/MetricsRecorder.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

final class MetricsRecorder {
    static let shared = MetricsRecorder()

    private static func makeDefaultSession() -> URLSession {
        let configuration = URLSessionConfiguration.ephemeral
        configuration.timeoutIntervalForRequest = 5
        configuration.timeoutIntervalForResource = 5
        return URLSession(configuration: configuration)
    }

    private static let defaultToggleProvider: () -> Bool = { ApexMediation.shared.metricsEnabled }

    private let queue = DispatchQueue(label: "com.rivalapexmediation.ctv.metrics", qos: .background)
    private var counters: [String: Int] = [:]
    private var latencies: [Double] = []
    private var flushTimer: DispatchSourceTimer?
    private var initialized = false
    private var config: SDKConfig?
    private let maxLatencySamples = 200
    private var flushInterval: TimeInterval = 30
    private var session: URLSession = MetricsRecorder.makeDefaultSession()
    private var toggleProvider: () -> Bool = MetricsRecorder.defaultToggleProvider

    private init() {}

    func initialize(config: SDKConfig, session: URLSession? = nil, flushInterval: TimeInterval = 30) {
        queue.async {
            self.config = config
            self.session = session ?? MetricsRecorder.makeDefaultSession()
            let envInterval = ProcessInfo.processInfo.environment["CTV_METRICS_FLUSH_SECONDS"].flatMap { Double($0) }
            let resolvedInterval = envInterval ?? flushInterval
            self.flushInterval = max(0.1, resolvedInterval)
            self.initialized = true
        }
    }

    func recordRequest(durationMs: Double, error: LoadError?) {
        queue.async {
            guard self.canCollect else { return }
            self.increment("requests_total")
            if error == nil || error == .noFill {
                self.increment("requests_success")
            } else {
                self.increment("requests_error")
                self.increment("requests_error_\(error?.reason ?? "unknown")")
            }
            self.addLatency(durationMs)
            self.scheduleFlushLocked()
        }
    }

    func recordPlayback(eventName: String) {
        guard !eventName.isEmpty else { return }
        queue.async {
            guard self.canCollect else { return }
            self.increment("playback_\(eventName)")
            self.scheduleFlushLocked()
        }
    }

    func recordTracker(eventName: String, success: Bool) {
        guard !eventName.isEmpty else { return }
        queue.async {
            guard self.canCollect else { return }
            let suffix = success ? "success" : "failure"
            self.increment("tracker_\(eventName)_\(suffix)")
            self.scheduleFlushLocked()
        }
    }

    private var canCollect: Bool {
        initialized && toggleProvider()
    }

    private func increment(_ key: String, delta: Int = 1) {
        counters[key, default: 0] += delta
    }

    private func addLatency(_ duration: Double) {
        guard duration > 0 else { return }
        latencies.append(duration)
        if latencies.count > maxLatencySamples {
            latencies.removeFirst(latencies.count - maxLatencySamples)
        }
    }

    private func scheduleFlushLocked() {
        guard flushTimer == nil else { return }
        let timer = DispatchSource.makeTimerSource(queue: queue)
        timer.schedule(deadline: .now() + flushInterval)
        timer.setEventHandler { [weak self] in
            self?.flushLocked()
        }
        timer.resume()
        flushTimer = timer
    }

    private func flushLocked() {
        flushTimer?.cancel()
        flushTimer = nil
        guard let cfg = config else {
            counters.removeAll()
            latencies.removeAll()
            return
        }
        guard canCollect else {
            counters.removeAll()
            latencies.removeAll()
            return
        }
        guard let payload = buildPayloadLocked(appId: cfg.appId),
              let url = URL(string: cfg.apiBaseUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/")) + "/sdk/metrics"),
              let body = try? JSONSerialization.data(withJSONObject: payload) else { return }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        if let key = cfg.apiKey {
            request.addValue("Bearer \(key)", forHTTPHeaderField: "Authorization")
        }
        request.httpBody = body
        session.dataTask(with: request).resume()
        counters.removeAll()
        latencies.removeAll()
    }

    private func buildPayloadLocked(appId: String) -> [String: Any]? {
        if counters.isEmpty && latencies.isEmpty { return nil }
        var payload: [String: Any] = [
            "timestamp": Int(Date().timeIntervalSince1970 * 1000),
            "sdk": "tvos",
            "appId": appId,
            "counters": counters
        ]
        if !latencies.isEmpty {
            let sorted = latencies.sorted()
            payload["request_latency_ms"] = [
                "p50": percentile(sorted, fraction: 0.50),
                "p95": percentile(sorted, fraction: 0.95),
                "p99": percentile(sorted, fraction: 0.99)
            ]
        }
        return payload
    }

    private func percentile(_ values: [Double], fraction: Double) -> Int {
        guard !values.isEmpty else { return 0 }
        let index = Int(Double(values.count - 1) * fraction)
        let clamped = max(0, min(values.count - 1, index))
        return Int(values[clamped])
    }

    func overrideToggleProvider(_ provider: (() -> Bool)?) {
        queue.async {
            self.toggleProvider = provider ?? MetricsRecorder.defaultToggleProvider
        }
    }

    func flushNowForTesting() {
        queue.async {
            self.flushLocked()
        }
    }

    func resetForTesting() {
        queue.sync {
            self.flushTimer?.cancel()
            self.flushTimer = nil
            self.counters.removeAll()
            self.latencies.removeAll()
            self.initialized = false
            self.config = nil
            self.flushInterval = 30
            self.session = MetricsRecorder.makeDefaultSession()
            self.toggleProvider = MetricsRecorder.defaultToggleProvider
        }
    }
}
